server/channels/app/post.go

```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
 	}
 
 	permalinkPreviewedPost := post.GetPreviewPost()
-	for _, cm := range channelMembers {
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
-		}
-
-		postForUser := a.sanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedPost, permalinkPreviewedChannel, cm.UserId)
+	if permalinkPreviewedPost != nil {
+		post.Metadata.Embeds[0].Data = permalinkPreviewedPost
+	} else {
+		return false, nil
```

Comment:
I think the various discussions below highlight some of the surprising behaviour, but to summarize, this method has implicit contracts with the caller (be sure to `.Add("post", postJSON)`) and undocumented side effects (populating `post.Metadata.Embeds[0].Data` if the user has access to it).

I wonder if most of this could just be resolved with a `publishWebsocketEventForPost` that encapsulates both the logic in this method and the fiddly bits that callers seem to have to juggle, e.g. all of this:
```go
	message.Add("post", postJSON)

	published, err := a.publishWebsocketEventForPermalinkPost(c, rpost, message)
	if err != nil {
		return nil, err
	}
	if !published {
		a.Publish(message)
	}
```

I'm 0/5 on whether we do this in this PR: just thinking out loud after reading this (existing) code for the first time in a while. I also want to balance the new complexity we're adding with the existing potential for confusion.

Reply:
Indeed, I agree. I will think some more on this, but I'd like to keep that refactor as a separate PR and focus this PR purely on the performance improvement.